### PR TITLE
Allow deep property matching via array syntax

### PR DESCRIPTION
**Pull request by captbaritone**
Inspired by `underscore-contrib`'s _.getPath` and lodash's deep property access, this adds a deep property access array shorthand to many aspects of Underscore, including:

* `_.property`
* `_.iteratee` and all methods that depend on it.

It also exposes a convenience function `_.get` for accessing deep properties with a default.

Possible Concerns
-----------------

While `_.property` does not get called in any place that would present a performance concern, `getLength` (which is created by `_.property`) gets used in many places which _may_ have performance ramifications.

I've placed the `isArray` call in the outer function of `_.property` in an attempt to reduce complexity inside the generated function. This required defining `isArray` earlier.

It's possible that this still isn't fast enough, and we should just define `getLength` manually.

Adds ~0.06kB (minified/gzipped)

Further work
------------

The following methods should probably also accept the array syntax:

* `_.result`
* `_.has`

The following companion methods should be created:

* `_.set`

Fixes: #2372, an elegant solution to an oft requested feature: (see: #2370, #2268, #1169, #1266)

### CODE REVIEW

File: underscore.js [hunk 1]
```diff
@@ -1279,9 +1288,7 @@
 
   // Is a given value an array?
   // Delegates to ECMA5's native Array.isArray
-  _.isArray = nativeIsArray || function(obj) {
-    return toString.call(obj) === '[object Array]';
-  };
+  _.isArray = isArray;
```
Comment (akre54) [inline underscore.js]: Can we remove or at least change the eslint rule? It's totally valid to lazily define methods on the `_` object provided they're not getting used right away.

Comment (captbaritone) [inline underscore.js]: Unfortunately this time it _is_ getting called right away. The outer `_.property` function (which uses `isArray`) gets called when we create `getLength`.

Comment (akre54) [inline underscore.js]: :/

What if we re-arranged the sections so that the `_.property` parts comes after `_.is[...]`?

File: underscore.js [hunk 2]
```diff
@@ -1446,6 +1453,13 @@
     return _.isFunction(value) ? value.call(object) : value;
   };
 
+  // Returns the, possibly deeply nested, value at `path` if defined.
+  // Otherwise, returns `fallback`.
+  _.get = function(object, path, fallback) {
+    var value = property(path)(object);
+    return value === void 0 ? fallback : value;
+  };
```
Comment (megawac) [inline underscore.js]: If we add `_.get` are we obliged to add `_.set`, `_.unset`?

Also you missed `_.result`

Comment (captbaritone) [inline underscore.js]: I mentioned `_.set`, `_.has` and `_.result` in the PR message as "Further Work". I wasn't sure if we wanted to tackle them all at once, or one at a time. (We could also pull `_.get` out of this PR if that would simplify things)

`_.unSet` was not on my radar.

To answer your question directly, I'm guessing we would be obliged, but I figured this PR would be a good forum to make that decision as a group.

Comment (akre54) [inline underscore.js]: I missed that in the commit message since I don't usually see the commit messages directly. We typically leave commit messages short (~1 sentence) and then write longer prose in the body of the Pull Request. Mind either copying that message to the pull request, or pointing out that there might be more text in the commit message in the future?

Comment (captbaritone) [inline underscore.js]: Hmm. Maybe I did something strange? I believe I posted (nearly) the same content in the commit and the pull request message. Here's what I see:

screen shot 2016-03-30 at 8 26 14 am

If that's not what you're seeing, let me know and I can try opening a fresh PR.

Comment (akre54) [inline underscore.js]: Oh no you're totally right. They just looked different to me. Carry on.

Comment (captbaritone) [inline underscore.js]: I'm going to delete these last few PR-message-related comments so that they don't distract from the topic at hand.

Comment (megawac) [inline underscore.js]: Okay, I'm aware you mentioned them, I didn't phrase my response well so apologies for that (was rushing to submit a paper). What I meant is what is the motivation and value of adding `_.get` (and committing to `_.set` and `_.unset`) rather than modifying `_.result` for paths.

Comment (captbaritone) [inline underscore.js]: I see. Yeah it make sense to start with `_.result` since it would be a superset of `_.get`'s functionality, and we can do it without committing to `_.set` and `_.unset` (We probably will need to update `_.has` as well).

Do you think we should aim for evaluating functions at every step a la https://github.com/lodash/lodash/commit/f65d36b3c1d87c95534e56fba98b576a452e8fb7 ?

**Comment by captbaritone** [comment]
Based on @megawac's feedback, I'm reworking this PR to:

1. Not include `_.get`
2. Add deep matching to `_.has` and `_.result`

It may be difficult to do so without adding some weight, so I figure we should do them all at once so we know what we're getting ourselves into.

File: underscore.js (hunk 2 above)
Comment (megawac) [inline underscore.js]: I'm not sure to be honest. I've never wanted that functionality personally, but the article that the commit you referenced was in response to was pretty convincing (ko.observables support)

**Comment by megawac** [comment]
Sounds good, for now I prefer that direction

File: test/functions.js [hunk 3]
```diff
@@ -694,6 +694,9 @@
       assert.deepEqual(_.toArray(cb(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)), _.range(1, 11));
     });
 
+    var deepProperty = _.iteratee(['a', 'b']);
+    assert.equal(deepProperty({a: {b: 2}}), 2, 'treats an array as a deep property accessor');
+
```
Comment (jdalton) [inline test/functions.js]: Compat note:
Lodash treats an array as shorthand for `matchesProperty`.

The `matchesProperty` shorthand is `[propName,propValue]` and as fast path for
the common case of `{propName:propValue}`


Not a biggie but something to be aware of.

Comment (captbaritone) [inline test/functions.js]: @jdalton Thanks for bringing this up. I'm a bit confused though. In lodash you use arrays as shorthand for a type of predicate which can already be expressed via an object (`_.matches`)?

I'm guessing from the phrase "fast path" and the perf graph you included that this somehow offers a performance gain? (Or maybe I'm misunderstanding?)

To me, deep property matching seems much more useful than a performance increase which users could presumably get in the places where performance was a concern by simply using `_.matchesProperty()` or their own optimized function directly.

Again, I could just be misinterpreting your comment, so feel free to set me straight. :smile:

File: underscore.js (hunk 1 above)
Comment (captbaritone) [inline underscore.js]: @akre54 I'm not sure I understand. If we move `_.property` below where the current `_.is[...]` methods live, it' won't be instantiated when we go to create `getLength`.

File: test/functions.js (hunk 3 above)
Comment (jdalton) [inline test/functions.js]: > In lodash you use arrays as shorthand for a type of predicate which can already be expressed via an object (`_.matches`)?

Yep. The `_.matches` shorthand is a bit heavier as it supports multiple properties.
You can think of `_.matches` implemented by way of `_.matchesProperty`. The `_.matches` method handles a collection of source properties to compare.

> I'm guessing from the phrase "fast path" and the perf graph you included that this somehow offers a performance gain?

Yep. It's a common enough case that a fast path was created.

> To me, deep property matching seems much more useful than a performance increase which users could presumably get in the places where performance was a concern by simply using `_.matchesProperty()` or their own optimized function directly.

In lodash `matches` and `matchesProperty` already perform partial deep matches of source objects:
```js
_.filter(objects, { 'a': { 'b': { 'c': 1 }, 'd': 2 } });
```
In Lodash the common case for deep property paths is strings in with arrays as a fallback for edge cases which is where `_.property(array)` would be used.
```js
_.filter(objects, 'a.b.c');
```

Comment (captbaritone) [inline test/functions.js]: Ah this all makes sense. Thanks for the clarification.

Comment (jridgewell) [inline test/functions.js]: lodash's `matchesProperty` is tripping a few people up, but I can't seem to find the issues. And I'm biased, because I dislike the `a.b.c` deep property syntax. `['a', 'b', 'c']` FTW.

Comment (jdalton) [inline test/functions.js]: I dig `a.b.c` but couldn't see supporting one without the other.
The array allows supporting symbols and other funky keys. I :heart: funk:

Comment (jdalton) [inline test/functions.js]: Folks wanting cross-lib interop can just use the explicit forms like `_.property`.

Comment (captbaritone) [inline test/functions.js]: Assuming Underscore is philosophically opposed to the `a.b.c` syntax, I think it makes most sense to treat arrays as deep property matching:

    // Underscore only (proposed)
    var adminPosts = _.filter(posts, ['author', 'isAdmin']);

    // Lodash only
    var adminPosts = _.filter(posts, 'author.isAdmin');
    var adminPosts = _.filter(posts, {author, {isAdmin: true}});

    // Compat (proposed)
    var adminPosts = _.filter(posts, _.property(['author', 'isAdmin']));

**Comment by captbaritone** [comment]
Closing in favor of #2494

File: test/functions.js (hunk 3 above)
Comment (jdalton) [inline test/functions.js]: I've [created a poll](https://twitter.com/jdalton/status/718165294678552576) as a gut check. Feel free to RT.

Comment (jdalton) [inline test/functions.js]: Results:

Comment (captbaritone) [inline test/functions.js]: The Tweet responses add some interesting color as well. Thanks for running that @jdalton. That being said, I don't think this thread is actually about (re)considering the string syntax. It seems pretty clear to me that the string syntax is not something Underscore wants to support.

Comment (jdalton) [inline test/functions.js]: Additional info at least lets them know they aren't inline with their users.
